### ISSUE TITLE
Pictures in recipes don't show up 

### ISSUE THREAD

**Issue by auanasgheps** [vabene1111/recipes#409]
### Version
Version: 0.14.2

### Bug description
Hi, I am using Recipes with Nginx which is serving media files.
I can see pictures in the main page, but not when I open a recipe.
![immagine](https://user-images.githubusercontent.com/20586878/108412971-b8d39900-722a-11eb-9aa8-f52f1f6e9a98.png)
![immagine](https://user-images.githubusercontent.com/20586878/108413031-cdb02c80-722a-11eb-94d9-ca959643b56e.png)

The browser is telling me that won't load the resource over http, but my server is being run in https. 

This is my nginx config for recipes

```
server {
    listen 8080 ssl;
    listen [::]:8080 ssl;

    server_name recipes.*;

    include /config/nginx/ssl.conf;

    client_max_body_size 0;

    # serve media files
    location /media {
        alias /recipes-media;
    }
	
    location / {
        include /config/nginx/proxy.conf;
        resolver 127.0.0.11 valid=30s;
        set $upstream_app recipes-app;
        set $upstream_port 8080;
        set $upstream_proto http;
        proxy_pass $upstream_proto://$upstream_app:$upstream_port;
		#proxy_set_header Host $http_host;
		proxy_set_header X-Forwarded-Proto https;
		
    }
}
```

**Comment by tdvantine** [comment]
Right click in the "image" in recipe view and paste the link it shows here. There was a similar [issue](https://github.com/vabene1111/recipes/issues/392), however according to the nginx conf you posted it shouldn't be related.

**Comment by auanasgheps** [comment]
Yes, is missing the port here as well!

**Comment by tdvantine** [comment]
I over looked it, you have it commented out `#proxy_set_header Host $http_host;` should be `proxy_set_header Host $http_host;`!

**Comment by vabene1111** [comment]
i kinda overlooked that issue with the update, everyone who uses the old install method will need to update the nginx config

**Comment by auanasgheps** [comment]
> I over looked it, you have it commented out `#proxy_set_header Host $http_host;` should be `proxy_set_header Host $http_host;`!

If I uncomment the option, the whole thing freaks out and I simply get **Bad Request (400)**
![immagine](https://user-images.githubusercontent.com/20586878/108423691-73b66380-7238-11eb-8996-0a2e593539c4.png)

**Comment by faspina** [comment]
Did you ever get the .conf file working for reverse project. I am using swag and I have the same problem above and can't figure it out. 

The change described didn't work , bad request 

In the main cookbook the url for the picture is  /media/recipes/ <media files>png
but in the recipe its the entire url in my case its the https://cookbook.mydomain.org/media/recipes/<media files> i have tried various settings for location/media but I can't get it to work

**Comment by vabene1111** [comment]
django rest determines the url for media files automatically according to the host headers it receives 

from what you posted above the two urls look the same or at least correct ? 
Can you post the base domain and full paths  from the search and the view ?

**Comment by faspina** [comment]
that is the full path teh cookbook page only has the /media/recipe path , but in the recipe is has the full url for the picture. 
 I got it to work with the swag/nginx setting below. 


`# make sure that your dns has a cname set for grafana and that your grafana container is not using a base url

server {
    listen 443 ssl;
    listen [::]:443 ssl;

    server_name cookbook.*;

    include /config/nginx/ssl.conf;


    client_max_body_size 16M;



	
	
    location / {
        include /config/nginx/proxy.conf;
        resolver 127.0.0.11 valid=30s;
        set $upstream_app cookbook;
        set $upstream_port 8080;
        set $upstream_proto http;
        proxy_pass $upstream_proto://$upstream_app:$upstream_port;
		
    }
    location /media/recipes {
        include /config/nginx/proxy.conf;
        resolver 127.0.0.11 valid=30s;
        set $upstream_app cookbook;
        set $upstream_port 8080;
        set $upstream_proto http;
        proxy_pass $upstream_proto://$upstream_app:$upstream_port;
		
    }
}`

**Comment by vabene1111** [comment]
that is great. I dont really know much about SWAG setups, do you think we should add something to the docs regarding this ? 

Is it just the nginx config that needs to be changed ?

**Comment by auanasgheps** [comment]
I am using swag too, which is simply a nginx container with let's encrypt and other stuff.
Either way 

> that is the full path teh cookbook page only has the /media/recipe path , but in the recipe is has the full url for the picture.
> I got it to work with the swag/nginx setting below.

Have you mounted /media/recipe inside swag container?

**Comment by faspina** [comment]
no maybe I should I have the SERVER MEDIA SET TO 1

**Comment by vabene1111** [comment]
serve media means that gunicorn serves your media files. This is most likely no problem for your setup but not recommended.

If the setting is 0 then you will need to point nginx to the mediafiles directory

**Comment by faspina** [comment]
Yep that was it , I had to setup a volumn 
container path: /recipes-media
host path: /mnt/user/appdata/recipes/media/

Then I put back the default recipes config from swag and that is when I realized the instruction that I completely ignored at the top. Got it working with server media turned off now. Thanks guys. Learn something new every day.  
```

## Version 2021/01/21
# make sure that your dns has a cname set for recipes
# make sure to mount /media/ in your swag container to point to your Recipes Media directory

# if using Authelia use this one:
# Doc: https://vabene1111.github.io/recipes/install/docker/#using-proxy-authentication

server {
    listen 443 ssl;
    listen [::]:443 ssl;

    server_name recipes.*;

    include /config/nginx/ssl.conf;

    client_max_body_size 0;

    # enable for ldap auth, fill in ldap details in ldap.conf
    #include /config/nginx/ldap.conf;

    # enable for Authelia
    #include /config/nginx/authelia-server.conf;
	
    # serve media files
    location /media/ {
        alias /recipes-media/;
    }
	
    location / {
        # enable the next two lines for http auth
        #auth_basic "Restricted";
        #auth_basic_user_file /config/nginx/.htpasswd;

        # enable the next two lines for ldap auth
        #auth_request /auth;
        #error_page 401 =200 /ldaplogin;

        # enable for Authelia
        #include /config/nginx/authelia-location.conf;

        include /config/nginx/proxy.conf;
        resolver 127.0.0.11 valid=30s;
        set $upstream_app recipes;
        set $upstream_port 8080;
        set $upstream_proto http;
        proxy_pass $upstream_proto://$upstream_app:$upstream_port;

    }
}
```

**Comment by auanasgheps** [comment]
My configuration is exactly the same as yours (see the first post), but does not work 100%. 

- serve media is set to 0 as recommended
- Ironically I mounted the same alias in the container

Pictures **do load** in the main page, but not when I open a recipe.

By inspecting the code, the website is trying to load a non-existent page.
The picture link is `http://recipes.example.com/media/recipes/fbb197d7-1f7e-4329-9d57-609a4d3c9f4a_17.png`
This link **does not exist** because the server is hosted at `https://recipes.example.com:8080`

HTTPS and custom port are missing!

I tried countless changes, but none of them has made a change.

**Comment by faspina** [comment]
Make the location the line for the media files the full path 

location /media/recipes

and sure media-recipes in swag is pointed to the recipe sub directory in the recipes docker 

My host path is actually  /mnt/user/appdata/recipes/media/recipes

**Comment by auanasgheps** [comment]
Just tested and does not make a difference
- Pictures in main page are still loaded (so paths are good)
- Picture in recipes are broken, still attempting to load without HTTPS and missing custom port.

**Comment by auanasgheps** [comment]
Below is the part of the code responsible to get the image.
I don't know the inner details, but from where it gets the URL?

I'm asking because in the homepage photos are loaded with a relative path, but it the recipe it uses the full (wrong) path. Why not use the same method? 

https://github.com/vabene1111/recipes/blob/e2b1115b3b3c27d02c773d4084205304be3224f4/vue/src/apps/RecipeView/RecipeView.vue#L105
https://github.com/vabene1111/recipes/blob/e2b1115b3b3c27d02c773d4084205304be3224f4/vue/src/apps/RecipeView/RecipeView.vue#L106

**Comment by vabene1111** [comment]
I am really not sure what you are doing with your setups but django dertermines the URL of the media files depending on the headers it receives with the api request.

Therefor you need to pass the proper headers to the gunicorn worker as configured in the default configuration file that is provided in the repository. I am not sure how you would need to change that to your setup but both the http vs https and the port issue are issues i came accorss during development and that were fixed in the nginx config

**Comment by albcp** [comment]
> Below is the part of the code responsible to get the image.
> I don't know the inner details, but from where it gets the URL?
> 
> I'm asking because, in the home page, photos are loaded with a relative path, but in the recipe it uses the full (wrong) path. Why not use the same method?
> 
> https://github.com/vabene1111/recipes/blob/e2b1115b3b3c27d02c773d4084205304be3224f4/vue/src/apps/RecipeView/RecipeView.vue#L105
> 
> https://github.com/vabene1111/recipes/blob/e2b1115b3b3c27d02c773d4084205304be3224f4/vue/src/apps/RecipeView/RecipeView.vue#L106

The image URL is included in the /api/recipe/<ID>/ response. It seems that the deserialization performed in api.py does not take the proxy_pass into account.

**Comment by vabene1111** [comment]
the thing is that during testing of the 0.14 beta we noticed this issue and thats why we added the proxy parameter to the nginx config

the generation of the url is different on the search and the recipe view page because one uses the api and the other uses django templates. The API uses django rest framework to generate the full linkts to the images. We noticed that ports would be ignored during and the protocol was wrong during testing but after adding the appropriate header/proxy parameters everthing worked. 

Are you sure the headers are actually going trough correctly ?

**Comment by auanasgheps** [comment]
I'm sure there's something wrong, because when I enable the recommended header
`proxy_set_header Host $http_host;`

I simply get Bad Request 400.

Are you referring to this one?

**Comment by vabene1111** [comment]
I think so. I have done so many things in the past weeks that i dont really remember which issue was related to which problem.

The Bad Request error might simply indicate that there is another problem. Could you provide me with some logs when you enable the header and do the request that causes the 400 error ?

**Comment by auanasgheps** [comment]
So when I enable the header, that's what the container says.
I've amended the real domain.

```
172.18.0.21 - - [23/Feb/2021:23:00:33 +0000] "GET /favicon.ico HTTP/1.1" 400 143 "https://recipes.example.com:8080/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
172.18.0.21 - - [23/Feb/2021:23:00:34 +0000] "GET / HTTP/1.1" 400 143 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
172.18.0.21 - - [23/Feb/2021:23:00:34 +0000] "GET /favicon.ico HTTP/1.1" 400 143 "https://recipes.example.com:8080/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
172.18.0.21 - - [23/Feb/2021:23:00:34 +0000] "GET /service-worker.js HTTP/1.1" 400 143 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
172.18.0.21 - - [23/Feb/2021:23:00:34 +0000] "GET / HTTP/1.1" 400 143 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
172.18.0.21 - - [23/Feb/2021:23:00:35 +0000] "GET /favicon.ico HTTP/1.1" 400 143 "https://recipes.example.com:8080/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
172.18.0.21 - - [23/Feb/2021:23:00:35 +0000] "GET /service-worker.js HTTP/1.1" 400 143 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
```

Here's what nginx says

```
192.168.0.100 - - [24/Feb/2021:00:00:33 +0100] "GET /service-worker.js HTTP/2.0" 400 143 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
192.168.0.100 - - [24/Feb/2021:00:00:33 +0100] "GET / HTTP/2.0" 400 143 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
192.168.0.100 - - [24/Feb/2021:00:00:33 +0100] "GET /favicon.ico HTTP/2.0" 400 143 "https://recipes.example.com:8080/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
192.168.0.100 - - [24/Feb/2021:00:00:34 +0100] "GET / HTTP/2.0" 400 143 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
192.168.0.100 - - [24/Feb/2021:00:00:34 +0100] "GET /favicon.ico HTTP/2.0" 400 143 "https://recipes.example.com:8080/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
192.168.0.100 - - [24/Feb/2021:00:00:34 +0100] "GET /service-worker.js HTTP/2.0" 400 143 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
192.168.0.100 - - [24/Feb/2021:00:00:34 +0100] "GET / HTTP/2.0" 400 143 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
192.168.0.100 - - [24/Feb/2021:00:00:35 +0100] "GET /favicon.ico HTTP/2.0" 400 143 "https://recipes.example.com:8080/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
192.168.0.100 - - [24/Feb/2021:00:00:35 +0100] "GET /service-worker.js HTTP/2.0" 400 143 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:85.0) Gecko/20100101 Firefox/85.0"
```

I don't think they say much. Can I get more verbose?

**Comment by vabene1111** [comment]
you can ty to set DEBUG=1 in the `.env` file, there should be some log somewhere that shows why the 400 is returned. 

What is you allowed host setting ? if it is set maybe try setting it back to `*`

**Comment by auanasgheps** [comment]
> What is you allowed host setting ? if it is set maybe try setting it back to `*`

Where is this option?

**Comment by vabene1111** [comment]
in the `.env` file the `ALLOWED_HOSTS` variable

**Comment by auanasgheps** [comment]
Ah yes, was always set to `*`.

I will enable the debug option and come back.

**Comment by auanasgheps** [comment]
That's what the page shows with DEBUG enabled.
Redacted my domain with "example.com"

```
Request Method: | GET
-- | --
http://recipes.example.com,recipes.example.com:8080/search/
3.1.7
DisallowedHost
Invalid HTTP_HOST header: 'recipes.example.com,recipes.example.com:8080'. The domain name provided is not valid according to RFC 1034/1035.
/opt/recipes/venv/lib/python3.8/site-packages/django/http/request.py, line 137, in get_host
/opt/recipes/venv/bin/python
3.8.8
['/opt/recipes',  '/opt/recipes/venv/bin',  '/usr/local/lib/python38.zip',  '/usr/local/lib/python3.8',  '/usr/local/lib/python3.8/lib-dynload',  '/opt/recipes/venv/lib/python3.8/site-packages']
Thu, 25 Feb 2021 11:38:59 +0100
```
There's also a lot of other information.

**Comment by vabene1111** [comment]
ok it appears that the host header that is passed on by your reverse proxy setup `recipes.example.com,recipes.example.com:8080` is not valid according to django / the RFC.

Is it possible that your setup adds multiple headers when you enable the header parameter above ? 

Something in the chain of services the request passes trough modifies the request header in a wrong way and causes this issue. If the request would get trough with the correct header it looks like everything would work fine.

**Comment by auanasgheps** [comment]
OK, update:

There are multiple headers passed because the "SWAG" config at:
`include /config/nginx/proxy.conf;`

Already includes
`proxy_set_header Host $host;`

So we are good, there's no need to explicitly specify it.
But the issue isn't solved.

I have to force the following parameter
`proxy_set_header X-Forwarded-Proto https;`

otherwise that's what I get when I try to login:
![image](https://user-images.githubusercontent.com/20586878/109143552-76e1b000-7760-11eb-867e-a0dff859c6c1.png)

**Comment by auanasgheps** [comment]
Another update:

If I add 
`proxy_set_header X-Forwarded-Proto $scheme;`
I can login and I don't get the CSRF error.

However, photos are still attempted to be loaded without HTTPS and without port.

**Comment by vabene1111** [comment]
Ok so from what i can figure this could be solved by adding the `CSRF_TRUSTED_ORIGINS` variable https://docs.djangoproject.com/en/3.1/ref/settings/#csrf-trusted-origins

but this would only solve the symptom not the actual problem. For some reason the Referer header does not match the Host header (at least thats what i can figure from the django docs, its not actually my area of most expertise). 

Have you seen the other issue that was taking about how to use SWAG ? To be honest i am a little lost on who has which exact problem at the moment but maybe there is something that could help you.

#429

**Comment by auanasgheps** [comment]
CSRF is just a side effect that I initially didn't know about. We can workaround it from nginx and from the app, although I tried to add `CSRF_TRUSTED_ORIGINS=recipes.example.com` to the env file and redeploy the container but does not make a difference.

But the main issue is always the same: cannot load photos in recipes pages, because the website is trying to load them without HTTPS and without the custom port.

**Comment by vabene1111** [comment]
`CSRF_TRUSTED_ORIGINS` wont do anything in .env as its not implemented yet. It was just a note for me that i could go that way. As that basically just disables a security measure that is there for a good reason i dont want to go that way.

I know that CSRF is just a side effect. I could most likely find a fix in the application but the problem does not lie there as far as i can tell. The django application and with it the djangor rest framework which renders the api dertmines its URL automatically based on the headers it receives. Since you somehow broke the headers with the SWAG container setup or something else in the chain it cant do so correctly.

The Problem is that i do not have enough insight into your setup to really help you here except that i am pretty sure that this is related to the headers

**Comment by auanasgheps** [comment]
SWAG is just a pre-configured container with nginx and other stuff.

All the configuration is public and found in [this ](https://github.com/linuxserver/docker-swag`)repository, along with the specific config for Recipes (which inherits some other settings)

There's a lot of people using this container, which is made to work with most of the software around, and I can see many people are using it with Recipes too.

I know nginx decently, but I have no knowledge of Django at all. From what I can see, the SWAG pre-configurations are pretty generic and should not break your app. Maybe Django is looking at for wrong header?

**Comment by vabene1111** [comment]
as django is used in probably millions of applications and so is the django rest framework that is responsible for the API i dont believe that the issue lies there either.

I would really love to help you further but i simply have no idea what causes your problem. I know that i had both the Port issue and the https issue and both were resolved by adding the correct headers. I also know that CSRF systems break if the host configuration is broken but that is were my knowledge of this ends. 

I know that the setup methods i provide as examples work and that people have working SWAG configurations. It saddens me to say but i dont know how to help you further.

**Comment by auanasgheps** [comment]
OK let's keep this open and see if we find a way. 

In the meantime are you planning to expose the CSRF variable so users are not that disrupted?

**Comment by vabene1111** [comment]
what do you mean by this ? the CSRF token is set as a cookie and forms contain it as a hidden field by default (as far as i remember).

Just as an idea: Why are you running Tandoor on port 8000 ? if you have a reverse proxy in place anyway and a domain (if you use Lets encrypt) shouldn't you be able to just use the setup that userers reportet was working in #429 ?

**Comment by auanasgheps** [comment]
Sorry I meant to implement `CSRF_TRUSTED_ORIGINS` that is supposed to fix that issue.

I am running the Tandoor container on port 8080 because is the default option, and my server is exposed over 8080 (just a coincidence). I don't understand the question 🤔

**Comment by kampsj** [comment]
I'm experiencing the same problem (running the default docker setup), albeit the other way around. Tandoor is trying to get the recipe image from a https location, while I'm running it without https.

**Comment by vabene1111** [comment]
although this might also be related to the headers you are using a completely different setup method. Could you open a new issue detailing your setup so i can see if anything looks wrong ?

**Comment by auanasgheps** [comment]
Reporting the fix from #542

Use the MEDIA_URL variable to specify the full path where medias are served, like
```
# If mediafiles are stored at a different location uncomment and change accordingly
MEDIA_URL=https://recipes.example.com:8080/media/
```

**Comment by vabene1111** [comment]
now i am feeling stupid for not figuring this out earlier 🙈  but at least its fixed i guess
